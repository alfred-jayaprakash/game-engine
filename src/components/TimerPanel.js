import React, {useState, useEffect} from 'react';
import {Row, Col, Progress} from 'reactstrap';

//
// TimerPanel. Pass the duration in secs as prop.duration
//
const TimerPanel = props => {
  const [timerVal, setTimerVal] = useState (100);
  const [timerStr, setTimerStr] = useState ('');
  const [timerCol, setTimerCol] = useState ('success');
  const {duration, currentRef, onTimeOver} = props;

  useEffect (
    () => {
      let currtime = parseInt (duration);
      setTimerStr (currtime + 's');
      //Create a timer
      const timer = setInterval (() => {
        currtime -= 1; //Reduce 1 sec
        setTimerStr (currtime + 's'); //Display the current sec
        setTimerVal (100 / parseInt (duration) * currtime); //Reduce the progess bar
        setTimerCol (
          currtime <= 10 ? (currtime < 5 ? 'danger' : 'warning') : 'success'
        );
        if (currtime === 0) {
          clearInterval (timer);
          onTimeOver (); //Call the onTimeOver function
        }
      }, 1000);

      return () => {
        //Cancel the timer
        clearInterval (timer);
      };
    },
    [duration, currentRef, onTimeOver]
  );

  return (
    <Row>
      <Col>
        <Progress bar color={timerCol} value={timerVal}>{timerStr}</Progress>
      </Col>
    </Row>
  );
};

export default TimerPanel;
