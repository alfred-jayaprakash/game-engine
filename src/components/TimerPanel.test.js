import React from 'react';
import {render, fireEvent, screen, act, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {shallow} from 'enzyme';
import TimerPanel from './TimerPanel';

let container;
let currentImage = '1.jpg';
let duration = 30;
let onTimeOver;

describe ('Snapshot tests', () => {
  beforeAll (() => {
    //ARRANGE
    onTimeOver = () => {};
  });

  it ('should render correctly without issues', () => {
    const component = shallow (
      <TimerPanel
        duration={duration}
        currentImage={currentImage}
        onTimeOver={onTimeOver}
      />
    );

    expect (component).toMatchSnapshot ();
  });
});

describe ('Functional tests', () => {
  beforeAll (() => {
    //ARRANGE
    onTimeOver = jest.fn ();
  });

  beforeEach (() => {
    container = document.createElement ('div');
    document.body.appendChild (container);
  });

  afterEach (() => {
    document.body.removeChild (container);
    container = null;
    onTimeOver.mockClear ();
  });

  it ('initial screen should load correctly', () => {
    render (
      <TimerPanel
        duration={duration}
        currentImage={currentImage}
        onTimeOver={onTimeOver}
      />,
      container
    );
  });
});
