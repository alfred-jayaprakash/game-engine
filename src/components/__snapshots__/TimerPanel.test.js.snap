// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests should render correctly without issues 1`] = `
<Row
  tag="div"
  widths={
    Array [
      "xs",
      "sm",
      "md",
      "lg",
      "xl",
    ]
  }
>
  <Col
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Progress
      bar={true}
      color="success"
      max={100}
      tag="div"
      value={100}
    />
  </Col>
</Row>
`;
